File: blender-addon/oh_my_blender/canonical.py
"""Canonical JSON serialization and revision hashing for Scene Snapshot v2.

Contract: docs/SCENE-SNAPSHOT-V2.md section 4. Must stay byte-identical to
packages/blender-director/src/canonical.ts; the shared fixture
packages/blender-director/test/fixtures/canonical-numbers.json guards parity
in both test suites. Pure stdlib so it runs outside Blender.
"""

from __future__ import annotations

import hashlib
import math
import unicodedata
from fractions import Fraction

TEN_POW_9 = 10**9
MAX_MAGNITUDE = 1e15

_ESCAPES = {
    '"': '\\"',
    "\\": "\\\\",
    "\b": "\\b",
    "\t": "\\t",
    "\n": "\\n",
    "\f": "\\f",
    "\r": "\\r",
}


def canonical_number(value: float | int) -> str:
    """Serialize per the spec: decimal rounded half-even to 1e-9.

    Trailing fractional zeros stripped, ``-0`` becomes ``"0"``, no exponent
    notation. The integer rule coincides with this algorithm for integral
    values, so all JSON numbers go through here.
    """
    if isinstance(value, bool):
        raise TypeError("bool is not a canonical number")
    if isinstance(value, float) and (math.isnan(value) or math.isinf(value)):
        raise ValueError(f"canonical number must be finite, got {value!r}")
    if abs(value) >= MAX_MAGNITUDE:
        raise ValueError(f"canonical number magnitude must be < 1e15, got {value!r}")
    # Fraction(float) is the exact binary64 value; round() on Fraction is
    # exact round-half-even. Rounding the magnitude matches the TypeScript
    # implementation, which rounds the (positive) mantissa expansion.
    scaled = round(Fraction(abs(value)) * TEN_POW_9)
    if scaled == 0:
        return "0"
    sign = "-" if value < 0 else ""
    integer_part, fraction_digits = divmod(scaled, TEN_POW_9)
    fraction_part = str(fraction_digits).rjust(9, "0").rstrip("0")
    if fraction_part:
        return f"{sign}{integer_part}.{fraction_part}"
    return f"{sign}{integer_part}"


def canonical_string(value: str) -> str:
    """NFC-normalize and serialize with JSON minimal escaping."""
    normalized = unicodedata.normalize("NFC", value)
    out = ['"']
    for character in normalized:
        escape = _ESCAPES.get(character)
        if escape is not None:
            out.append(escape)
        elif ord(character) < 0x20:
            out.append(f"\\u{ord(character):04x}")
        else:
            out.append(character)
    out.append('"')
    return "".join(out)


def canonical_json(value: object) -> str:
    """Serialize a parsed JSON value to canonical text.

    Code-point-sorted NFC keys, no whitespace, numbers per canonical_number.
    Array order is preserved; semantic sorting happens in the schema layer.
    """
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return canonical_number(value)
    if isinstance(value, str):
        return canonical_string(value)
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(canonical_json(item) for item in value) + "]"
    if isinstance(value, dict):
        entries = sorted(
            ((unicodedata.normalize("NFC", key), nested) for key, nested in value.items()),
            key=lambda entry: entry[0],
        )
        return "{" + ",".join(f"{canonical_string(key)}:{canonical_json(nested)}" for key, nested in entries) + "}"
    raise TypeError(f"value is not canonical JSON serializable: {type(value).__name__}")


def canonical_revision(value: object) -> str:
    """Lowercase-hex SHA-256 of the UTF-8 canonical bytes."""
    return hashlib.sha256(canonical_json(value).encode("utf-8")).hexdigest()

File: packages/director-core/src/canonical.ts
// Canonical JSON serialization and revision hashing for Scene Snapshot v2.
// Contract: docs/SCENE-SNAPSHOT-V2.md §4. Must stay byte-identical to
// blender-addon/cclay/canonical.py; the shared fixture
// test/fixtures/canonical-numbers.json guards parity in both test suites.

import { createHash } from "node:crypto";

const TEN_POW_9 = 1_000_000_000n;
const MAX_MAGNITUDE = 1e15;

/**
 * Serialize a finite binary64 as decimal rounded half-even to 1e-9,
 * trailing fractional zeros stripped, `-0` as `"0"`, no exponent notation.
 *
 * The integer rule of the spec (base-10, no leading zeros) coincides with
 * this algorithm for integral values, so all JSON numbers go through here.
 */
export function canonicalNumber(value: number): string {
	if (!Number.isFinite(value)) throw new Error(`canonical number must be finite, got ${value}`);
	if (Math.abs(value) >= MAX_MAGNITUDE) throw new Error(`canonical number magnitude must be < 1e15, got ${value}`);
	const view = new DataView(new ArrayBuffer(8));
	view.setFloat64(0, value);
	const bits = view.getBigUint64(0);
	if (bits << 1n === 0n) return "0"; // +0 and -0
	const negative = bits >> 63n === 1n;
	const exponentBits = Number((bits >> 52n) & 0x7ffn);
	const mantissaBits = bits & 0xf_ffff_ffff_ffffn;
	// Exact magnitude is mantissa * 2^exponent.
	const mantissa = exponentBits === 0 ? mantissaBits : mantissaBits | (1n << 52n);
	const exponent = (exponentBits === 0 ? 1 : exponentBits) - 1075;
	// scaled = round_half_even(magnitude * 1e9), computed exactly over BigInt.
	let scaled: bigint;
	if (exponent >= 0) {
		scaled = (mantissa * TEN_POW_9) << BigInt(exponent);
	} else {
		const numerator = mantissa * TEN_POW_9;
		const denominator = 1n << BigInt(-exponent);
		const quotient = numerator / denominator;
		const remainder = numerator % denominator;
		const doubled = remainder * 2n;
		scaled = doubled > denominator || (doubled === denominator && (quotient & 1n) === 1n) ? quotient + 1n : quotient;
	}
	if (scaled === 0n) return "0";
	const sign = negative ? "-" : "";
	const integerPart = scaled / TEN_POW_9;
	const fractionPart = (scaled % TEN_POW_9).toString().padStart(9, "0").replace(/0+$/, "");
	return fractionPart === "" ? `${sign}${integerPart}` : `${sign}${integerPart}.${fractionPart}`;
}

/** NFC-normalize and serialize with JSON minimal escaping. */
export function canonicalString(value: string): string {
	const normalized = value.normalize("NFC");
	let out = '"';
	for (const character of normalized) {
		const code = character.codePointAt(0) as number;
		if (character === '"') out += '\\"';
		else if (character === "\\") out += "\\\\";
		else if (code < 0x20) {
			if (character === "\b") out += "\\b";
			else if (character === "\t") out += "\\t";
			else if (character === "\n") out += "\\n";
			else if (character === "\f") out += "\\f";
			else if (character === "\r") out += "\\r";
			else out += `\\u${code.toString(16).padStart(4, "0")}`;
		} else out += character;
	}
	return `${out}"`;
}

function compareCodePoints(left: string, right: string): number {
	const leftPoints = [...left];
	const rightPoints = [...right];
	const shared = Math.min(leftPoints.length, rightPoints.length);
	for (let index = 0; index < shared; index++) {
		const difference = (leftPoints[index].codePointAt(0) as number) - (rightPoints[index].codePointAt(0) as number);
		if (difference !== 0) return difference;
	}
	return leftPoints.length - rightPoints.length;
}

/**
 * Serialize a parsed JSON value to canonical bytes: code-point-sorted NFC keys,
 * no whitespace, numbers per canonicalNumber. Array order is preserved —
 * semantic sorting happens in the schema layer before serialization.
 */
export function canonicalJson(value: unknown): string {
	if (value === null) return "null";
	if (typeof value === "boolean") return value ? "true" : "false";
	if (typeof value === "number") return canonicalNumber(value);
	if (typeof value === "string") return canonicalString(value);
	if (Array.isArray(value)) return `[${value.map(canonicalJson).join(",")}]`;
	if (typeof value === "object") {
		const entries = Object.entries(value as Record<string, unknown>)
			.map(([key, nested]) => [key.normalize("NFC"), nested] as const)
			.sort(([left], [right]) => compareCodePoints(left, right));
		return `{${entries.map(([key, nested]) => `${canonicalString(key)}:${canonicalJson(nested)}`).join(",")}}`;
	}
	throw new Error(`value is not canonical JSON serializable: ${typeof value}`);
}

/** Lowercase-hex SHA-256 of the UTF-8 canonical bytes. */
export function canonicalRevision(value: unknown): string {
	return createHash("sha256")
		.update(Buffer.from(canonicalJson(value), "utf8"))
		.digest("hex");
}
